File: runners/js/leetcode_runner.js
const fs = require('fs');
const path = require('path');

// TreeNode definition
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// ListNode definition
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

// Make TreeNode and ListNode available globally so user solutions can reference them
global.TreeNode = TreeNode;
global.ListNode = ListNode;

function parseValue(raw, type) {
    raw = raw.trim();
    switch (type) {
        case 'int': return parseInt(raw, 10);
        case 'double': return parseFloat(raw);
        case 'boolean': return raw.toLowerCase() === 'true';
        case 'String': return raw.startsWith('"') ? JSON.parse(raw) : raw;
        case 'int[]':
        case 'List<Integer>':
        case 'int[][]':
        case 'List<List<Integer>>':
        case 'String[]':
        case 'List<String>':
            return JSON.parse(raw);
        case 'TreeNode': return buildTree(JSON.parse(raw));
        case 'ListNode': return buildList(JSON.parse(raw));
        default: throw new Error(`Unsupported type: ${type}`);
    }
}

function serializeValue(value, type) {
    if (value === null || value === undefined) return 'null';
    switch (type) {
        case 'int':
        case 'double':
        case 'boolean':
            return String(value);
        case 'String':
            return JSON.stringify(value);
        case 'int[]':
        case 'List<Integer>':
        case 'int[][]':
        case 'List<List<Integer>>':
        case 'String[]':
        case 'List<String>':
            return JSON.stringify(value);
        case 'TreeNode': return serializeTree(value);
        case 'ListNode': return serializeList(value);
        default: return String(value);
    }
}

function buildTree(arr) {
    if (!arr || !arr.length || arr[0] === null) return null;
    const root = new TreeNode(arr[0]);
    const queue = [root];
    let i = 1;
    while (queue.length && i < arr.length) {
        const node = queue.shift();
        if (i < arr.length && arr[i] !== null) {
            node.left = new TreeNode(arr[i]);
            queue.push(node.left);
        }
        i++;
        if (i < arr.length && arr[i] !== null) {
            node.right = new TreeNode(arr[i]);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

function serializeTree(root) {
    if (!root) return '[]';
    const result = [];
    const queue = [root];
    while (queue.length) {
        const node = queue.shift();
        if (node) {
            result.push(node.val);
            queue.push(node.left);
            queue.push(node.right);
        } else {
            result.push(null);
        }
    }
    while (result.length && result[result.length - 1] === null) result.pop();
    return '[' + result.map(v => v === null ? 'null' : String(v)).join(',') + ']';
}

function buildList(arr) {
    if (!arr || !arr.length) return null;
    const head = new ListNode(arr[0]);
    let curr = head;
    for (let i = 1; i < arr.length; i++) {
        curr.next = new ListNode(arr[i]);
        curr = curr.next;
    }
    return head;
}

function serializeList(head) {
    if (!head) return '[]';
    const vals = [];
    let curr = head;
    while (curr) {
        vals.push(curr.val);
        curr = curr.next;
    }
    return '[' + vals.join(',') + ']';
}

// Main
const problemDir = process.argv[2];
if (!problemDir) {
    console.error('Usage: leetcode_runner.js <problem-directory>');
    process.exit(1);
}

const metadata = JSON.parse(fs.readFileSync(path.join(problemDir, 'metadata.json'), 'utf8'));
const inputLines = fs.readFileSync(path.join(problemDir, 'input.txt'), 'utf8')
    .split('\n')
    .filter(line => line.trim() !== '');

const Solution = require(path.join(path.resolve(problemDir), 'solution.js'));
const sol = new Solution();
const method = sol[metadata.method].bind(sol);

const args = metadata.params.map((param, i) => {
    const line = i < inputLines.length ? inputLines[i] : '';
    return parseValue(line, param.type);
});

const result = method(...args);
console.log(serializeValue(result, metadata.return));

File: runners/python/leetcode_runner.py
import sys
import os
import json
import importlib.util
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def parse_value(raw, type_str):
    raw = raw.strip()
    if type_str == "int":
        return int(raw)
    elif type_str == "double":
        return float(raw)
    elif type_str == "boolean":
        return raw.lower() == "true"
    elif type_str == "String":
        return json.loads(raw) if raw.startswith('"') else raw
    elif type_str in ("int[]", "List<Integer>"):
        return json.loads(raw)
    elif type_str in ("int[][]", "List<List<Integer>>"):
        return json.loads(raw)
    elif type_str in ("String[]", "List<String>"):
        return json.loads(raw)
    elif type_str == "TreeNode":
        return build_tree(json.loads(raw))
    elif type_str == "ListNode":
        return build_list(json.loads(raw))
    else:
        raise ValueError(f"Unsupported type: {type_str}")


def serialize_value(value, type_str):
    if value is None:
        return "null"
    if type_str in ("int", "double"):
        return str(value)
    elif type_str == "boolean":
        return "true" if value else "false"
    elif type_str == "String":
        return json.dumps(value)
    elif type_str in ("int[]", "List<Integer>", "int[][]", "List<List<Integer>>",
                       "String[]", "List<String>"):
        return json.dumps(value, separators=(",", ":"))
    elif type_str == "TreeNode":
        return serialize_tree(value)
    elif type_str == "ListNode":
        return serialize_list(value)
    else:
        return str(value)


def build_tree(arr):
    if not arr or arr[0] is None:
        return None
    root = TreeNode(arr[0])
    queue = deque([root])
    i = 1
    while queue and i < len(arr):
        node = queue.popleft()
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1
    return root


def serialize_tree(root):
    if not root:
        return "[]"
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(str(node.val))
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append("null")
    while result and result[-1] == "null":
        result.pop()
    return "[" + ",".join(result) + "]"


def build_list(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    curr = head
    for v in arr[1:]:
        curr.next = ListNode(v)
        curr = curr.next
    return head


def serialize_list(head):
    if not head:
        return "[]"
    vals = []
    curr = head
    while curr:
        vals.append(str(curr.val))
        curr = curr.next
    return "[" + ",".join(vals) + "]"


def main():
    if len(sys.argv) < 2:
        print("Usage: leetcode_runner.py <problem-directory>", file=sys.stderr)
        sys.exit(1)

    problem_dir = sys.argv[1]

    with open(os.path.join(problem_dir, "metadata.json")) as f:
        metadata = json.load(f)

    method_name = metadata["method"]
    params = metadata["params"]
    return_type = metadata["return"]

    with open(os.path.join(problem_dir, "input.txt")) as f:
        input_lines = [line.rstrip("\n") for line in f.readlines()]

    # Import solution module
    solution_path = os.path.join(problem_dir, "solution.py")
    spec = importlib.util.spec_from_file_location("solution", solution_path)
    mod = importlib.util.module_from_spec(spec)

    # Inject TreeNode and ListNode into the module's namespace before executing
    mod.TreeNode = TreeNode
    mod.ListNode = ListNode
    spec.loader.exec_module(mod)

    sol = mod.Solution()
    method = getattr(sol, method_name)

    arguments = []
    for i, param in enumerate(params):
        line = input_lines[i] if i < len(input_lines) else ""
        arguments.append(parse_value(line, param["type"]))

    result = method(*arguments)
    print(serialize_value(result, return_type))


if __name__ == "__main__":
    main()
